File: 백준/Gold/7662. 이중 우선순위 큐/이중 우선순위 큐.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(br.readLine());

        while (t-- > 0) {
            int k = Integer.parseInt(br.readLine());
            PriorityQueue<Node> minHeap = new PriorityQueue<>();
            PriorityQueue<Node> maxHeap = new PriorityQueue<>(Collections.reverseOrder());
            boolean[] visited = new boolean[k];

            for (int i = 0; i < k; i++) {
                String[] input = br.readLine().split(" ");
                char cmd = input[0].charAt(0);
                int n = Integer.parseInt(input[1]);

                if (cmd == 'I') {
                    Node node = new Node(n, i);
                    minHeap.offer(node);
                    maxHeap.offer(node);
                    visited[i] = true;
                } else {
                    if (n == 1) {
                        while (!maxHeap.isEmpty() && !visited[maxHeap.peek().index]) {
                            maxHeap.poll();
                        }
                        if (!maxHeap.isEmpty()) {
                            visited[maxHeap.poll().index] = false;
                        }
                    } else {
                        while (!minHeap.isEmpty() && !visited[minHeap.peek().index]) {
                            minHeap.poll();
                        }
                        if (!minHeap.isEmpty()) {
                            visited[minHeap.poll().index] = false;
                        }
                    }
                }
            }

            while (!minHeap.isEmpty() && !visited[minHeap.peek().index]) {
                minHeap.poll();
            }
            while (!maxHeap.isEmpty() && !visited[maxHeap.peek().index]) {
                maxHeap.poll();
            }

            if (!minHeap.isEmpty() && !maxHeap.isEmpty()) {
                System.out.println(maxHeap.peek().value + " " + minHeap.peek().value);
            } else {
                System.out.println("EMPTY");
            }
        }
    }

    static class Node implements Comparable<Node> {
        int value, index;

        Node(int value, int index) {
            this.value = value;
            this.index = index;
        }

        @Override
        public int compareTo(Node other) {
            return Integer.compare(this.value, other.value);
        }
    }
}

File: 백준/Gold/7662. 이중 우선순위 큐/이중 우선순위 큐.py
import heapq
import sys
input = sys.stdin.readline

t = int(input())
for _ in range(t):
    k = int(input())
    min_heap, max_heap = [], []
    visited = [False] * k 

    for i in range(k):
        cmd, n = input().split()
        n = int(n)
        if cmd == 'I':  # 삽입
            heapq.heappush(min_heap, (n, i))
            heapq.heappush(max_heap, (-n, i))
            visited[i] = True
        else:  # 삭제
            if n == 1:  # 최대값 삭제
                while max_heap and not visited[max_heap[0][1]]:
                    heapq.heappop(max_heap)
                if max_heap:
                    visited[max_heap[0][1]] = False
                    heapq.heappop(max_heap)
            else:  # 최소값 삭제
                while min_heap and not visited[min_heap[0][1]]:
                    heapq.heappop(min_heap)
                if min_heap:
                    visited[min_heap[0][1]] = False
                    heapq.heappop(min_heap)

    # 남은 최소/최대 정리
    while min_heap and not visited[min_heap[0][1]]:
        heapq.heappop(min_heap)
    while max_heap and not visited[max_heap[0][1]]:
        heapq.heappop(max_heap)

    if min_heap and max_heap:
        print(-max_heap[0][0], min_heap[0][0])
    else:
        print("EMPTY")
